### PR TITLE
Added basic gating as TimeSeries method

### PR DESCRIPTION
**Pull request by tjma12**
This PR adds gating as a `TimeSeries` class method. The method whitens the time series, marks all points above a threshold, clusters them within a time window, and applies inverse Planck windows to zero out the data. The **unwhitened**, gated time series is returned.

An example can be found here (behind LIGO.org auth): https://ldas-jobs.ligo.caltech.edu/~thomas.massinger/notebook/demonstrate-gwpy-gating.html

Whitening arguments can be passed to the underlying whitening used to locate gating points.

### CODE REVIEW

**Comment by alurban** [comment]
@duncanmmacleod, the build failure reported by Travis seems to be from `pycbc` trying to import `glue.segments`

File: gwpy/timeseries/timeseries.py [hunk 1]
```diff
@@ -58,6 +58,45 @@ def _fft_length_default(dt):
     return int(max(2, numpy.ceil(2048 * dt.decompose().value)))
 
 
+def _discover_gate_points(ts, thresh, window):
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: Is this method a generic 'points above threshold' detector? If so, can we move it into the `Series` class as an instance method and give it an appropriate name?

File: gwpy/timeseries/timeseries.py [hunk 2]
```diff
@@ -1567,6 +1606,98 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
...
+        >>> ax.set_xlim(1225542713, 1225542733)
+        >>> ax.legend()
+        >>> overlay.show()
+        """
+        # Find points to gate based on a threshold
+        if whiten:
+            gates = _discover_gate_points(self.whiten(**whiten_kwargs),
+                                          threshold, cluster_window)
+        else:
+            gates = _discover_gate_points(self, threshold, cluster_window)
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: I would modify this to:

```python
        data = self.whiten(**whiten_kwargs) if whiten else self
        gates = _discover_gate_points(data, threshold, cluster_window)
```

File: gwpy/timeseries/timeseries.py [hunk 3]
```diff
@@ -1567,6 +1606,98 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
...
+        out = self.copy()
+
+        # Iterate over list of indices to gate and apply each one
+        for gate in gates:
+            nzero = int(abs(tzero)*self.sample_rate.value)
+            npad = int(abs(tpad)*self.sample_rate.value)
+            half = nzero + npad
+            ntotal = 2*half
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: it looks like all of this is gate-independent, so could be moved outside the loop.

File: gwpy/timeseries/timeseries.py [hunk 4]
```diff
@@ -1567,6 +1606,98 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
...
+            left_idx_window = half - (gate - left_idx)
+            right_idx_window = half + (right_idx - gate)
+
+            window = 1 - planck(ntotal, nleft=npad, nright=npad)
+            window = window[left_idx_window:right_idx_window]
+            window_timeseries = TimeSeries(numpy.ones_like(self.value),
+                                           dt=self.dt, t0=self.t0)
+            window_timeseries[left_idx:right_idx] = window
+
+            out *= window_timeseries
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: Can you instead just do this and not have to create another timeseries?:

```python
            out[left_idx:right_idx] *= window
```

**Comment by duncanmmacleod** [comment]
@tjma12, I have worked around the pycbc failure in #1053, please rebase.

File: gwpy/timeseries/timeseries.py (hunk 1 above)
Comment (tjma12) [reply gwpy/timeseries/timeseries.py]: It's a "points above threshold" detector with the automatic clustering that you'd want for gating. If I move it into the `Series` class, would you like me to make the clustering optional?

**Comment by tjma12** [comment]
> @tjma12, I have worked around the pycbc failure in #1053, please rebase.

Rebased and squashed commits, I'll implement the other fixes once the build succeeds

File: gwpy/timeseries/timeseries.py (hunk 1 above)
Comment (tjma12) [reply gwpy/timeseries/timeseries.py]: Will replace this with `scipy.signal.find_peaks`

File: gwpy/timeseries/timeseries.py [hunk 5]
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
         out = in_.convolve(tdw, window=window)
         return out * numpy.sqrt(2 * in_.dt.decompose().value)
 
+    def gate(self, tzero=1.0, tpad=0.5, whiten=True,
+             threshold=50., cluster_window=0.5, **whiten_kwargs):
+        """Discovers a set of indices where gating windows should be placed
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: I don't think this docstring is correct. Perhaps something like

```python
        """Mask clustered peaks above threshold using planck windows
 
        Parameters
        ------------
        ...
```

File: gwpy/timeseries/timeseries.py [hunk 6]
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
...
+            a copy of the original `TimeSeries` that has had gating windows
+            applied
+
+        Examples
+        --------
+
+        Read data into a `TimeSeries`
+        >>> from gwpy.timeseries import TimeSeries
+        >>> data = TimeSeries.get('H1:GDS-CALIB_STRAIN', 1225542623,
+                                   1225542823)
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: Can you modify this example to use `fetch_open_data`?

File: gwpy/timeseries/timeseries.py [hunk 7]
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
...
+                              fftlength=4, overlap=2, method='median')
+
+        Plot the original data and the gated data
+        >>> overlay = data.plot(dpi=150, label='Ungated',
+                                color='dodgerblue', zorder=3)
+        >>> ax = overlay.gca()
+        >>> ax.plot(gated, label='Gated', color='orange', zorder=2)
+        >>> ax.set_xlim(1225542713, 1225542733)
+        >>> ax.legend()
+        >>> overlay.show()
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: Can you post in this ticket an example of the figure you get?

File: gwpy/timeseries/timeseries.py [hunk 8]
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
...
+        """
+        # Find points to gate based on a threshold
+        data = self.whiten(**whiten_kwargs) if whiten else self
+        window_samples = cluster_window*data.sample_rate.value
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]:
```suggestion
        window_samples = cluster_window * data.sample_rate.value
```

File: gwpy/timeseries/timeseries.py [hunk 9]
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
...
+        gates = signal.find_peaks(data.value, height=threshold,
+                                  distance=window_samples)[0]
+        out = self.copy()
+
+        # Iterate over list of indices to gate and apply each one
+        nzero = int(abs(tzero)*self.sample_rate.value)
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]:
```suggestion
        nzero = int(abs(tzero) * self.sample_rate.value)
```

File: gwpy/timeseries/timeseries.py [hunk 10]
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
...
+        npad = int(abs(tpad)*self.sample_rate.value)
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]:
```suggestion
        npad = int(abs(tpad) * self.sample_rate.value)
```

File: gwpy/timeseries/timeseries.py [hunk 11]
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
...
+        half = nzero + npad
+        ntotal = 2*half
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]:
```suggestion
        ntotal = 2 * half
```

**Comment by duncanmmacleod** [comment]
@tjma12, what's the status of this PR? It looks from @alurban's comment that you are going to put in one or more unit tests, which would be great.

**Comment by tjma12** [comment]
@duncanmmacleod I plan to address your requested changes and write up a unit test today

File: gwpy/timeseries/timeseries.py (hunk 6 above)
Comment (tjma12) [reply gwpy/timeseries/timeseries.py]: Done, switched this to use a large glitch from O1 open data

File: gwpy/timeseries/timeseries.py (hunk 7 above)
Comment (tjma12) [reply gwpy/timeseries/timeseries.py]: See attached figure that is produced by updated docstring

gating_example

**Comment by tjma12** [comment]
I wrapped the data in an `abs()` when running the peak finding since whitened data can glitch negatively. Unit test is next.

**Comment by duncanmmacleod** [comment]
@tjma12, it turns out that scipy.signal.find_peaks` is new in scipy-1.1.0, I don't think we should change it, but can you add some importerror protectionn so that users get told what is wrong when they try to run `TimeSeries.gate` and it doesn't work.

You can auto-skip the unit test using a decorator

```python
from ...testing.utils import skip_minimum_version

...

        @skip_minimum_version("scipy", "1.1.0")
        def test_gate(self, ...)
```

**Comment by alurban** [comment]
@tjma12, what @duncanmmacleod said, I was going to suggest a direct import like this:

```python
try:
    from scipy.signal import find_peaks
except ImportError:
    raise ImportError('find_peaks is only available in scipy >= 1.1.0')
# Find points to gate based on a threshold
data = self.whiten(**whiten_kwargs) if whiten else self
window_samples = cluster_window * data.sample_rate.value
gates = find_peaks(abs(data.value), height=threshold,
                   distance=window_samples)[0]
```

File: gwpy/timeseries/timeseries.py [hunk 12]
```diff
@@ -1626,6 +1626,12 @@ def gate(self, tzero=1.0, tpad=0.5, whiten=True,
         >>> ax.legend()
         >>> overlay.show()
         """
+        try:
+             from scipy.signal import find_peaks
+        except ImportError:
+             raise ImportError("Must have scipy>=1.1.0 to utilize "
```
Comment (duncanmmacleod) [inline gwpy/timeseries/timeseries.py]: @tjma12, can you implement it a little different, to preserve the traceback:

```python
        except ImportError as exc:
            exc.args = ("scipy.signal.find_peaks is only available in scipy >= 1.1.0",)
            raise
````

Comment (tjma12) [reply gwpy/timeseries/timeseries.py]: Both have been implemented

**Pull request merged**